### PR TITLE
JUCX: Active Messages API.

### PR DESCRIPTION
**Pull request by petro-rudenko**
## What
 Active Messages API support in JUCX

## How ?
UcpAmData  class wrapper that incapsulates logic for eager and rndv active message data.

### CODE REVIEW

File: bindings/java/src/main/java/org/openucx/jucx/ucp/UcpAmData.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        this.ucpAmRecvCallback = ucpAmRecvCallback;
+        this.address = address;
+        this.length = length;
+        this.flags = flags;
+    }
+
+    @Override
+    public String toString() {
+        return "UcpAmData{" +
+            "address=" + address +
```
Comment (yosefe) [inline bindings/java/src/main/java/org/openucx/jucx/ucp/UcpAmData.java:36]: hex

File: bindings/java/src/main/java/org/openucx/jucx/ucp/UcpAmData.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            ", length=" + length +
+            ", received=" + isReceived() +
+            '}';
+    }
+
+    /**
+     * Whether actual data is received or need to call {@link UcpAmData#receive(long, UcxCallback)}
+     */
+    public boolean isReceived() {
```
Comment (yosefe) [inline bindings/java/src/main/java/org/openucx/jucx/ucp/UcpAmData.java:45]: isDataValid

File: bindings/java/src/main/java/org/openucx/jucx/ucp/UcpAmRecvCallback.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+ * {@link UcpWorker#progress()} is not allowed. It is recommended to define
+ * callbacks with relatively short execution time to avoid blocking of
+ * communication progress.
+ */
+public abstract class UcpAmRecvCallback {
+    UcpWorker worker;
+
+    void setWorker(UcpWorker worker) {
+        this.worker = worker;
+    }
```
Comment (yosefe) [inline bindings/java/src/main/java/org/openucx/jucx/ucp/UcpAmRecvCallback.java:21]: maybe hide this from API and save only on UcpAmData

File: bindings/java/src/main/java/org/openucx/jucx/ucp/UcpWorker.java [hunk 4]
```diff
@@ -55,8 +63,56 @@ public UcpListener newListener(UcpListenerParams params) {
...
+     */
+    public void setAmRecvHandler(int amId, UcpAmRecvCallback callback) {
+        if (callback != null) {
+            amRecvHandlers.put(amId, callback);
+            callback.setWorker(this);
+        } else {
+            amRecvHandlers.remove(amId);
+        }
+        setAmRecvHandlerNative(getNativeId(), amId, callback);
+    }
```
Comment (yosefe) [inline bindings/java/src/main/java/org/openucx/jucx/ucp/UcpWorker.java:93]: split to add/remove

File: bindings/java/src/main/native/jucx_common_def.cc [hunk 5]
```diff
@@ -45,10 +52,15 @@ extern "C" JNIEXPORT jint JNICALL JNI_OnLoad(JavaVM *jvm, void* reserved) {
     jclass jucx_callback_cls = env->FindClass("org/openucx/jucx/UcxCallback");
     jclass ucp_rkey_cls_local = env->FindClass("org/openucx/jucx/ucp/UcpRemoteKey");
     jclass ucp_tag_msg_cls_local = env->FindClass("org/openucx/jucx/ucp/UcpTagMessage");
+    jclass jucx_endpoint_cls_local = env->FindClass("org/openucx/jucx/ucp/UcpEndpoint");
+    jclass jucx_am_data_cls_local = env->FindClass("org/openucx/jucx/ucp/UcpAmData");
+    jclass am_recv_handler = env->FindClass("org/openucx/jucx/ucp/UcpAmRecvCallback");
```
Comment (yosefe) [inline bindings/java/src/main/native/jucx_common_def.cc:57]: am_recv_callback_cls_local

File: bindings/java/src/main/native/worker.cc [hunk 6]
```diff
@@ -63,7 +63,21 @@ Java_org_openucx_jucx_ucp_UcpWorker_createWorkerNative(JNIEnv *env, jclass cls,
...
     if (status != UCS_OK) {
         JNU_ThrowExceptionByStatus(env, status);
+        return -1L;
     }
+
+    ucp_worker_attr_t attr = {0};
+    attr.field_mask = UCP_WORKER_ATTR_FIELD_MAX_AM_HEADER;
+
+    status = ucp_worker_query(ucp_worker, &attr);
+
```
Comment (yosefe) [inline bindings/java/src/main/native/worker.cc:73]: remove space line

File: bindings/java/src/test/java/org/openucx/jucx/UcpEndpointTest.java [hunk 7]
```diff
@@ -594,4 +596,107 @@ public void onError(int ucsStatus, String errorMsg) {
...
+                    assertTrue(request.isCompleted());
+                }
+            });
+
+        requests[1] = worker2.recvTaggedNonBlocking(recvHeader, null);
+        requests[4] = ep.sendAmNonBlocking(1, 0L, 0L,
+            UcxUtils.getAddress(data), dataSize, UcpConstants.UCP_AM_SEND_FLAG_EAGER, null);
+
+
+        while (!Arrays.stream(requests).allMatch(r -> (r != null) && (r.isCompleted()))) {
```
Comment (yosefe) [inline bindings/java/src/test/java/org/openucx/jucx/UcpEndpointTest.java:685]: extra () in r.isCompleted()

**Review by yosefe — COMMENTED** [review]
run through clang format

File: bindings/java/src/main/java/org/openucx/jucx/ucp/UcpAmData.java (hunk 1 above)
Comment (yosefe) [reply bindings/java/src/main/java/org/openucx/jucx/ucp/UcpAmData.java:35]: unhandled

File: bindings/java/src/main/java/org/openucx/jucx/ucp/UcpWorker.java [hunk 8]
```diff
@@ -83,13 +83,23 @@ public long getMaxAmHeaderSize() {
...
-            amRecvHandlers.put(amId, callback);
-            callback.setWorker(this);
-        } else {
-            amRecvHandlers.remove(amId);
+        if (callback == null) {
+            removeAmRecvHandler(amId);
+            return;
         }
-        setAmRecvHandlerNative(getNativeId(), amId, callback);
+        Object[] callbackAndWorker = new Object[2];
```
Comment (yosefe) [inline bindings/java/src/main/java/org/openucx/jucx/ucp/UcpWorker.java:90]: maybe use Pair?

Comment (petro-rudenko) [reply bindings/java/src/main/java/org/openucx/jucx/ucp/UcpWorker.java:90]: There's no such class in java. To not create one just for this case, decided just to use an array of 2 elements

**Review by yosefe — APPROVED** [review]
pls squash

**Pull request merged**